File: node/src/config/firebase.ts
import admin from 'firebase-admin';
import { Logger } from '../utils/logger';
import * as fs from 'fs';
import * as path from 'path';

let firebaseApp: admin.app.App;
let db: admin.firestore.Firestore;

/**
 * Initialize Firebase Admin SDK
 */
export function initializeFirebase(): void {
  try {
    const projectId = process.env.FIREBASE_PROJECT_ID;
    const credentialsPath = process.env.FIREBASE_CREDENTIALS_PATH;

    if (!projectId) {
      throw new Error('FIREBASE_PROJECT_ID is not set in environment variables');
    }

    // Check if running in emulator mode
    const emulatorHost = process.env.FIREBASE_EMULATOR_HOST;

    if (emulatorHost) {
      // Use Firebase emulator for testing
      Logger.log(`Connecting to Firebase emulator at ${emulatorHost}`);
      process.env.FIRESTORE_EMULATOR_HOST = emulatorHost;

      firebaseApp = admin.initializeApp({
        projectId,
      });
    } else {
      // Production/development mode - use service account
      if (!credentialsPath) {
        throw new Error('FIREBASE_CREDENTIALS_PATH is not set in environment variables');
      }

      const absolutePath = path.resolve(credentialsPath);

      if (!fs.existsSync(absolutePath)) {
        throw new Error(`Firebase credentials file not found at: ${absolutePath}`);
      }

      const serviceAccount = JSON.parse(fs.readFileSync(absolutePath, 'utf8'));

      firebaseApp = admin.initializeApp({
        credential: admin.credential.cert(serviceAccount),
        projectId,
      });

      Logger.log('Firebase Admin SDK initialized successfully');
    }

    db = admin.firestore();
    const databaseId = process.env.FIREBASE_DATABASE_ID || '(default)';

    Logger.log(`Connected to Firestore database: ${databaseId}`);
  } catch (error) {
    Logger.error('Failed to initialize Firebase', error);
    throw error;
  }
}

/**
 * Get Firestore instance
 */
export function getFirestore(): admin.firestore.Firestore {
  if (!db) {
    throw new Error('Firestore not initialized. Call initializeFirebase() first.');
  }
  return db;
}

/**
 * Get Firebase app instance
 */
export function getFirebaseApp(): admin.app.App {
  if (!firebaseApp) {
    throw new Error('Firebase not initialized. Call initializeFirebase() first.');
  }
  return firebaseApp;
}

/**
 * Get Firestore server timestamp
 */
export function getServerTimestamp(): admin.firestore.FieldValue {
  return admin.firestore.FieldValue.serverTimestamp();
}

/**
 * Helper to get bracket collection reference
 */
export function getBracketCollection(eventId: string) {
  return db.collection(`event/${eventId}/brackets`);
}

/**
 * Helper to get dispute collection reference
 */
export function getDisputeCollection(eventId: string) {
  return db.collection(`event/${eventId}/disputes`);
}

export default {
  initializeFirebase,
  getFirestore,
  getFirebaseApp,
  getServerTimestamp,
  getBracketCollection,
  getDisputeCollection
};

File: cloud_api/app/core/firebase.py
"""
Firebase Admin SDK initialization.
Ported from node/src/config/firebase.ts
"""
import os
import firebase_admin
from firebase_admin import credentials, firestore
from app.core.config import settings
from loguru import logger


_firestore_client = None


def initialize_firebase():
    """
    Initialize Firebase Admin SDK.
    Supports both production (with credentials file) and emulator mode.
    """
    global _firestore_client

    try:
        # Configure emulator if specified
        if settings.FIREBASE_EMULATOR_HOST:
            os.environ["FIRESTORE_EMULATOR_HOST"] = settings.FIREBASE_EMULATOR_HOST
            logger.info(f"Using Firebase emulator at {settings.FIREBASE_EMULATOR_HOST}")

        # Initialize Firebase Admin
        if settings.FIREBASE_CREDENTIALS_PATH and os.path.exists(settings.FIREBASE_CREDENTIALS_PATH):
            cred = credentials.Certificate(settings.FIREBASE_CREDENTIALS_PATH)
            firebase_admin.initialize_app(cred, {
                'projectId': settings.FIREBASE_PROJECT_ID,
            })
            logger.info(f"Firebase initialized with credentials from {settings.FIREBASE_CREDENTIALS_PATH}")
        else:
            # Initialize without credentials (uses default application credentials)
            firebase_admin.initialize_app(options={
                'projectId': settings.FIREBASE_PROJECT_ID,
            })
            logger.info("Firebase initialized with default application credentials")

        # Get Firestore client
        _firestore_client = firestore.client()
        logger.info("Firestore client initialized")

        return _firestore_client

    except Exception as e:
        logger.error(f"Failed to initialize Firebase: {e}")
        raise


def get_firestore_client():
    """Get the Firestore client instance."""
    global _firestore_client
    if _firestore_client is None:
        _firestore_client = initialize_firebase()
    return _firestore_client


# Initialize on module import
try:
    db = get_firestore_client()
except Exception as e:
    logger.warning(f"Firebase initialization deferred: {e}")
    db = None
